program/src/processor.rs

```diff
@@ -76,30 +76,38 @@ fn get_total_rewards_checked(
         return Err(PaladinRewardsError::IncorrectHolderRewardsPoolAddress.into());
     }
 
-    let holder_rewards_pool_data = holder_rewards_pool_info.try_borrow_data()?;
-    let holder_rewards_pool_state =
-        bytemuck::try_from_bytes::<HolderRewardsPool>(&holder_rewards_pool_data)
-            .map_err(|_| ProgramError::InvalidAccountData)?;
-
-    Ok(holder_rewards_pool_state.total_rewards)
+    Ok(())
 }
 
-fn calculate_reward_share(
-    token_supply: u64,
-    token_account_balance: u64,
-    total_rewards: u64,
-) -> Result<u64, ProgramError> {
+fn calculate_rewards_per_token(rewards: u64, token_supply: u64) -> Result<u128, ProgramError> {
     if token_supply == 0 {
         return Ok(0);
     }
-    // Calculation: (token_amount / total_token_supply) * pool_rewards
+    // Calculation: rewards / token_supply
     //
-    // However, multiplication is done first to avoid truncation, ie:
-    // (token_amount * pool_rewards) / total_token_supply
-    (token_account_balance as u128)
-        .checked_mul(total_rewards as u128)
+    // Scaled by 1e9 to store 9 decimal places of precision.
+    (rewards as u128)
+        .checked_mul(1_000_000_000)
         .and_then(|product| product.checked_div(token_supply as u128))
-        .and_then(|share| u64::try_from(share).ok())
+        .ok_or(ProgramError::ArithmeticOverflow)
+}
+
+fn calculate_eligible_rewards(
+    current_rewards_per_token: u128,
+    last_rewards_per_token: u128,
+    token_account_balance: u64,
+) -> Result<u64, ProgramError> {
+    // Calculation: (current_rewards_per_token - last_rewards_per_token) *
+    // token_account_balance
+    let marginal_rate = current_rewards_per_token.saturating_sub(last_rewards_per_token);
+    if marginal_rate == 0 {
+        return Ok(0);
+    }
```

Comment:
Normally, I would say that this should never happen, but since we're now using imprecise math, I wonder if this might actually fail.

As a future test, we should make sure there's no attack where someone can distribute 1 lamport and potentially make rewards per token go down. If that passes, then we can harden this logic

Reply:
One of the tests does use a single lamport for the reward.

https://github.com/paladin-bladesmith/rewards-program/blob/e594c345f72dda9c73e10e2e8f9d1c37c8cbf54e/program/tests/distribute_rewards.rs#L298-L310

Maybe I can just add another case where there's a starting `rewards_per_token` rate?

Reply:
Actually, never mind, even with a tiny distribution, the number should never go down -- you'll just add 0. Ie. if there are 10^10 tokens, and 1 lamport is distributed, then the accumulated rate just stays the same.

In that case, this should be able to use checked math